### PR TITLE
👤🏷️ Guest mentions 

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] Requires https://github.com/nextcloud/server/pull/16331

### Notes
* `@"guest/<sha1(webrtc session id)>"` is used to mention a guest
* After reloading the chat, the mentions are not for "you" anymore. This is because we don't reuse the session id, so you get a new one (see #1793 )

### Todo
- [x] ~~Get guest mentions from comments~~ =>  https://github.com/nextcloud/server/pull/16331
- [x] Parse mentions and create the rich object parameters
- [x] Render the guest mentions in the chat view
- [x] Make sure the autocomplete uses the guest avatar
- [x] ~~Think about browser notifications~~ => https://github.com/nextcloud/spreed/issues/1975

### CODE REVIEW

**Comment by nickvergessen** [comment]
> Mentions to a guest are not highlighted when seen by that guest (isCurrentUser is not set to true).

This works pretty fine here, did you see the 2nd Note? I guess that is what you ran into.

**Comment by danxuliu** [comment]
> did you see the 2nd Note? I guess that is what you ran into.

I did, that is not the issue; the problem seems to be that sometimes [`sessionId` is not yet set when setting the room](https://github.com/nextcloud/spreed/pull/1974/files#diff-c74d955c6532055f49d56e992a7101caR88) (probably related to the external signaling server, as I made a quick test without it and in that case it worked).

**Comment by nickvergessen** [comment]
Integration tests added 

However I can't check the external signaling server.
For code point of view, this here should make sure it only joins the room when a session was retrieved:
https://github.com/nextcloud/spreed/blob/71f66fe9b1e86fb244f9c7809bdf27c4c68e3d17/js/signaling.js#L1028-L1037

But maybe you can investigate and fix it?

**Comment by danxuliu** [comment]
>  this here should make sure it only joins the room when a session was retrieved

Related: https://github.com/nextcloud/spreed/issues/1896

> But maybe you can investigate and fix it?

At least, I will try :-P

**Comment by danxuliu** [comment]
Rebased onto master to add acceptance tests that needed #2018.

I am still working on the issue with the external signaling server.

File: js/views/chatview.js [hunk 1]
```diff
@@ -603,14 +611,15 @@
...
-				if ($element.data('user-id')) {
+				if ($element.data('user-id') && $element.data('user-id').substr(0, 6) !== 'guest/') {
 					if ($element.data('user-id') === 'all') {
 						$element.addClass('avatar icon icon-contacts');
 					} else {
 						$element.avatar($element.data('user-id'), size, undefined, false, undefined, $element.data('user-display-name'));
 					}
 				} else {
-					$element.imageplaceholder('?', $element.data('displayname'), size);
+					var firstChar = $element.data('user-display-name') ? $element.data('user-display-name').substr(0, 1) : 'G';
```
Comment (danxuliu) [inline js/views/chatview.js]: Why the change from _?_ to _G_ for guests without a name set?

Comment (nickvergessen) [reply js/views/chatview.js]: We used G already for the guest at the chat input and in the participant list. So now the chat messages are consistent with that

Comment (nickvergessen) [reply js/views/chatview.js]: Hmm at least that is what Jan told me, but it seems it's not the case everywhere, let me check.

Comment (danxuliu) [reply js/views/chatview.js]: No, the participant list and the avatar next to where the guest sets his name both use _?_; the chat messages, on the other hand, use _G_.

Comment (nickvergessen) [reply js/views/chatview.js]: Jan-C. Borchardt (16th July, 12:06)
> Also some stuff regarding guests:
> - Marco is shown with an M (in grey circle) in the right sidebar participant list, but with a "?" in grey circle in the chat. Should be an M in both cases, and also can be colored based on placeholder?
> - Also, would it be possible to @mention guests?

Comment (nickvergessen) [reply js/views/chatview.js]: So if a name is set we should use that letter, otherwise `G` or `?` @jancborchardt

Comment (nickvergessen) [reply js/views/chatview.js]: Reverted back to `?` as default, but it uses the first char, when a name is set.

**Comment by nickvergessen** [comment]
> * A guest that has never set her name is shown with the G avatar instead of ? in the chat view
* A guest that has set her name is shown with the ? avatar instead of the first letter in the mention autocomplete

Fixed

> A guest that has cleared her name is shown with the ? avatar but no name at all (not even Guest) in the chat view.

Can not reproduce

**Comment by nickvergessen** [comment]
Fixed

File: js/views/chatview.js (hunk 1 above)
Comment (jancborchardt) [reply js/views/chatview.js]: > Reverted back to `?` as default, but it uses the first char, when a name is set.

Yep, that’s perfect! Also then it should be grey when no name is set, but colored as soon as a name is set, is that possible?

Comment (nickvergessen) [reply js/views/chatview.js]: That is possible, but I'd like to quickly tell the difference between guests and normal users. Also having the guest avatar colored, would allow to impersonate users that have not uploaded an avatar, because your guest name will generate the same avatar.

Comment (jancborchardt) [reply js/views/chatview.js]: Maybe it’s good then if we always append ` (guest)` to the name then, or at least in the participant list like we do it with ` (moderator)`?

Cause most people in the world don’t have Nextcloud and will likely be guests – making that look nice too would be cool. :) (Also in the future it would also be cool to have avatar upload for guests etc. – ref Allow guest users to set avatar #357)

**Comment by nickvergessen** [comment]
What's the status here, the external signaling still has the issue?

**Comment by danxuliu** [comment]
I have workarounded the issue with the external signaling server by generating a hashed session ID attribute in the model whenever the session ID changes and using that attribute from the chat view, instead of setting the hashed session ID in the chat view when the room was set and not changing it again.

This still has the issue that during the first moments that a guest joins the room the highlighting of mentions to the guest still do not work, but they will (for newer messages) as soon as the room is fetched again with the proper session ID.

I think that I have the proper fix (which should also fix some other issues as well :-D ), but I would like to test it better first, so something for later. Anyway, I think that with the workaround this pull request is in a good enough shape to be merged ;-)

**Comment by nickvergessen** [comment]
> This still has the issue that during the first moments that a guest joins the room the highlighting of mentions to the guest still do not work, but they will (for newer messages) as soon as the room is fetched again with the proper session ID.

Sounds good enough for now.
